pkg/combustion/kubernetes.go

```diff
@@ -122,28 +138,116 @@ func configureRKE2(ctx *image.Context) (string, error) {
 	return rke2InstallScript, nil
 }
 
-func copyKubernetesConfig(ctx *image.Context, distro string) (string, error) {
+func parseKubernetesConfig(ctx *image.Context) (map[string]any, error) {
+	auditDefaultCNI := func() {
+		auditMessage := fmt.Sprintf("Kubernetes CNI not explicitly set, defaulting to: %s", cniDefaultValue)
+		log.Audit(auditMessage)
+	}
+
+	config := map[string]any{}
+
 	if !isComponentConfigured(ctx, k8sConfigDir) {
-		zap.S().Info("Kubernetes config file not provided")
-		return "", nil
+		auditDefaultCNI()
+		zap.S().Infof("Kubernetes config file not provided, proceeding with CNI: %s", cniDefaultValue)
+
+		config[cniKey] = cniDefaultValue
+		return config, nil
 	}
 
 	configDir := generateComponentPath(ctx, k8sConfigDir)
 	configFile := filepath.Join(configDir, k8sConfigFile)
 
-	_, err := os.Stat(configFile)
+	b, err := os.ReadFile(configFile)
 	if err != nil {
 		if errors.Is(err, fs.ErrNotExist) {
-			return "", fmt.Errorf("kubernetes component directory exists but does not contain config.yaml")
+			return nil, fmt.Errorf("kubernetes component directory exists but does not contain config.yaml")
+		}
+		return nil, fmt.Errorf("reading kubernetes config file: %w", err)
+	}
+
+	if err = yaml.Unmarshal(b, &config); err != nil {
+		return nil, fmt.Errorf("parsing kubernetes config file: %w", err)
+	}
+
+	if _, ok := config[cniKey]; !ok {
+		auditDefaultCNI()
+		zap.S().Infof("CNI not set in config file, proceeding with CNI: %s", cniDefaultValue)
+
+		config[cniKey] = cniDefaultValue
+	}
+
+	return config, nil
+}
+
+func extractCNI(config map[string]any) (cni string, multusEnabled bool, err error) {
+	switch configuredCNI := config[cniKey].(type) {
+	case string:
+		if configuredCNI == "" {
+			return "", false, fmt.Errorf("cni not configured")
 		}
-		return "", fmt.Errorf("error checking kubernetes config file: %w", err)
+
+		var cnis []string
+		for _, cni = range strings.Split(configuredCNI, ",") {
+			cnis = append(cnis, strings.TrimSpace(cni))
+		}
+
+		return parseCNIs(cnis)
+
+	case []string:
+		return parseCNIs(configuredCNI)
+
+	case []any:
+		var cnis []string
+		for _, cni := range configuredCNI {
+			c, ok := cni.(string)
+			if !ok {
+				return "", false, fmt.Errorf("invalid cni value: %v", cni)
+			}
+			cnis = append(cnis, c)
+		}
+
+		return parseCNIs(cnis)
+
+	default:
+		return "", false, fmt.Errorf("invalid cni: %v", configuredCNI)
+	}
+}
+
+func parseCNIs(cnis []string) (cni string, multusEnabled bool, err error) {
```

Comment:
The validations here basically replicate the [RKE2 behaviour](https://github.com/rancher/rke2/blob/master/pkg/cli/cmds/server.go#L169). We discussed this with @rdoxenham and we'd want to establish the same rules that the particular distribution enforces and not silently fix mistakes on behalf of the user.

I believe we should be able to extract this out to the new validation framework eventually.